### ISSUE TITLE
HIDPI displays cause tiny UI and tiny text

### ISSUE THREAD

**Issue by triclops200** [atlas-engineer/next#293]
There should be a way for next to either detect hidpi displays and autozoom or at least a config option to scale everything by some factor. Currently, it's almost unusable on my 4k 15" laptop screen

**Comment by Ambrevar** [comment]
Sorry for taking to so long to reply to this, it fell under my radar.

You are talking about scaling the UI (e.g. minibuffer) right?  What about the web page content itself?

@jmercouris @vindarel: I think scaling the UI is just a matter of passing the right font size to `cl-markup`.  We should then write a wrapper function to make sure it's used systematically.  Thoughts?

**Comment by jmercouris** [comment]
This shall work for the Minibuffer :) For content within the web view I think we'd need to set a "default zoom factor" or something like that. 

So to summarize, I see two components:

1. Detect/Set Scaling Factor
2. Scale Minibuffer Contents
3. Zoom/Scale Main Window contents

Thoughts on the above?

**Comment by Ambrevar** [comment]
How do other web browsers deal about the content with HIDPI, any idea?

**Comment by Ambrevar** [comment]
Thanks for the details.
I'll work on it.

**Comment by danielsz** [comment]
Same problem here. It's possible to zoom in for the browser content, but I don't know what to do with the minibuffer, which is tiny.

**Comment by Ambrevar** [comment]
Currently the minibuffer (which is just another web view) does not set
the font size, it uses the default.  See `minibuffer.lisp` if you are
interested.
Patches welcome :)

A minibuffer overhaul is coming soon, stay tuned!

**Comment by Ambrevar** [comment]
Since ffaae9c we can set the default zoom level and minibuffer font:

```lisp
(in-package :next)

(defun my-buffer-defaults (buffer)
  (setf (zoom-ratio-default buffer) 1.1)
  (unzoom-page :buffer buffer))          ; Needed for non-web-mode buffers.

(defun my-minibuffer-defaults (minibuffer)
  (setf
   (max-lines minibuffer) 6
   (minibuffer-line-height minibuffer) "1.1em"
   (minibuffer-font-size minibuffer) "1.1em"))

(defun my-interface-defaults ()
  (hooks:add-to-hook (hooks:object-hook *interface* 'buffer-make-hook)
                     #'my-buffer-defaults)
  (hooks:add-to-hook (hooks:object-hook *interface* 'minibuffer-make-hook)
                     #'my-minibuffer-defaults))

(hooks:add-to-hook '*after-init-hook* #'my-interface-defaults)
```

Also see #151.

**Comment by Ambrevar** [comment]
Feel free to re-open if there is anything else.

**Comment by danielsz** [comment]
Running the code provided above yields the following error. What am I missing? Thank you.
```
Could not load the lisp init file /home/arch/daniel/.config/next/init.lisp:
The hook NEXT::*AFTER-INIT-HOOK* could not be found.
```

**Comment by Ambrevar** [comment]
You need to update next to 1.3.3.

**Comment by Drainful** [comment]
Having used Next on a hi dpi screen, I have found a 2 problems that arise when operating at high zoom levels:

1. SCROLL-TO-TOP and SCROLL-TO-BOTTOM only take you a portion of the way across the page when at high zoom.
2. When at high zoom and scrolled down on a page, link hints above a certain height are squished into one point. The greater the distance you have scrolled down the page, the lower the threshold. The included screenshots show this page at 2.0 zoom on a high dpi screen at increasing scroll amounts.

- https://ibb.co/1KP0nHZ -- top of page
- https://ibb.co/JdwR927 -- slightly scrolled, some hints are missing
- https://ibb.co/jT09fnx -- scrolled further, more hints are missing
- https://ibb.co/sKBN0HS -- below this point there are no link hints

I was able to hack the scrolling functions to scale their distance based on zoom (https://gist.github.com/Drainful/dce0e5dcbb62e334e1e14a360b36f1fd.js), but I don't really understand the parenscript involved in displaying link hints, so I'm not sure how to fix that problem.

**Comment by Ambrevar** [comment]
Hi Adrian,

Adrian Fullmer <notifications@github.com> writes:

> Having used Next on a hi dpi screen, I have found a 2 problems that arise when operating at high zoom levels:
>
> 1. SCROLL-TO-TOP and SCROLL-TO-BOTTOM only take you a portion of the way across the page when at high zoom.
> 2. When at high zoom and scrolled down on a page, link hints above a certain height are squished into one point. The greater the distance you have scrolled down the page, the lower the threshold. The included screenshots show this page at 2.0 zoom on a high dpi screen at increasing scroll amounts.
>
> - https://ibb.co/1KP0nHZ -- top of page
> - https://ibb.co/JdwR927 -- slightly scrolled, some hints are missing
> - https://ibb.co/jT09fnx -- scrolled further, more hints are missing
> - https://ibb.co/sKBN0HS -- below this point there are no link hints
>
> I was able to hack the scrolling functions to scale their distance based on zoom (https://gist.github.com/Drainful/dce0e5dcbb62e334e1e14a360b36f1fd.js), but I don't really understand the parenscript involved in displaying link hints, so I'm not sure how to fix that problem.

weird, I can't see your message on GitHub anymore... :(

Anyways, a possible answer to this could come after we re-architecture
Next and have full bindings for the web renderers, which would allow us
to use the the latter for zooming.  This should be solve the zooming all
at once.
